crates/currency/src/lib.rs

```diff
@@ -14,23 +15,419 @@ use frame_support::{
 use orml_traits::{MultiCurrency, MultiReservableCurrency};
 use pallet_transaction_payment::OnChargeTransaction;
 use sp_runtime::{
-    traits::{AtLeast32BitUnsigned, DispatchInfoOf, MaybeSerializeDeserialize, PostDispatchInfoOf, Saturating, Zero},
+    traits::{
+        AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, DispatchInfoOf,
+        MaybeSerializeDeserialize, PostDispatchInfoOf, Saturating, UniqueSaturatedInto, Zero,
+    },
     transaction_validity::InvalidTransaction,
+    FixedPointNumber, FixedPointOperand,
 };
-use sp_std::{fmt::Debug, marker::PhantomData};
+use sp_std::{
+    convert::{TryFrom, TryInto},
+    fmt::Debug,
+    marker::PhantomData,
+};
+
+pub use monetary::Amount;
+pub use pallet::*;
+
+mod types;
+pub use types::CurrencyConversion;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::pallet_prelude::*;
+
+    /// ## Configuration
+    /// The pallet's configuration trait.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + orml_tokens::Config<Balance = BalanceOf<Self>> {
+        type UnsignedFixedPoint: FixedPointNumber<Inner = BalanceOf<Self>>
+            + Encode
+            + EncodeLike
+            + Decode
+            + MaybeSerializeDeserialize;
+
+        type SignedInner: Debug
+            + CheckedDiv
+            + TryFrom<BalanceOf<Self>>
+            + TryInto<BalanceOf<Self>>
+            + MaybeSerializeDeserialize;
+
+        type SignedFixedPoint: FixedPointNumber<Inner = SignedInner<Self>>
+            + Encode
+            + EncodeLike
+            + Decode
+            + MaybeSerializeDeserialize;
+
+        type Balance: AtLeast32BitUnsigned
+            + FixedPointOperand
+            + MaybeSerializeDeserialize
+            + FullCodec
+            + Copy
+            + Default
+            + Debug;
+
+        /// Wrapped currency: INTERBTC.
+        #[pallet::constant]
+        type GetWrappedCurrencyId: Get<CurrencyId<Self>>;
+
+        type CurrencyConversion: types::CurrencyConversion<Amount<Self>, CurrencyId<Self>>;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        ArithmeticOverflow,
+        ArithmeticUnderflow,
+        TryIntoIntError,
+        InvalidCurrency,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
```

Comment:
Perhaps I'm missing something, but why does this need to be a pallet if there are no storage items?

Reply:
For the error type

Reply:
We don't need a pallet for the `Error` type, look at the `bitcoin` crate for example

Reply:
But then you need error conversions everywhere, plus you lose information that the error occurred in the `currency` crate